File: ts/src/adapters.ts
import type { FileSystem } from "./fs/operations.js";
import type { ToolResult } from "./tools.js";
import { tools, callTool } from "./tools.js";

// OpenAI / OpenAI-compatible (Groq, Together, Fireworks, etc.)
export function openai(fs: FileSystem) {
  const toolDefs = tools.map(t => ({
    type: "function" as const,
    function: { name: t.name, description: t.description, parameters: t.parameters },
  }));

  async function handleToolCall(name: string, args: string | Record<string, unknown>): Promise<ToolResult> {
    const parsed = typeof args === "string" ? JSON.parse(args) : args;
    return callTool(fs, name, parsed);
  }

  return { tools: toolDefs, handleToolCall };
}

// Anthropic
export function anthropic(fs: FileSystem) {
  const toolDefs = tools.map(t => ({
    name: t.name,
    description: t.description,
    input_schema: t.parameters as Record<string, unknown>,
  }));

  async function handleToolCall(name: string, input: Record<string, unknown>): Promise<ToolResult> {
    return callTool(fs, name, input);
  }

  return { tools: toolDefs, handleToolCall };
}

File: python/agent_vfs/adapters.py
from __future__ import annotations

import json
from typing import Any

from .operations import FileSystem
from .tools import TOOLS, call_tool


def openai(fs: FileSystem):
    """Create OpenAI-compatible tool definitions and handler.

    Usage:
        tools, handle = openai(fs)
        response = client.chat.completions.create(model="gpt-4o", messages=messages, tools=tools)
        for call in response.choices[0].message.tool_calls:
            result = handle(call.function.name, call.function.arguments)
    """
    tool_defs = [
        {
            "type": "function",
            "function": {
                "name": t["name"],
                "description": t["description"],
                "parameters": t["parameters"],
            },
        }
        for t in TOOLS
    ]

    def handle_tool_call(name: str, args: str | dict[str, Any]) -> dict:
        parsed = json.loads(args) if isinstance(args, str) else args
        return call_tool(fs, name, parsed)

    return tool_defs, handle_tool_call


def anthropic(fs: FileSystem):
    """Create Anthropic-compatible tool definitions and handler.

    Usage:
        tools, handle = anthropic(fs)
        response = client.messages.create(model="claude-sonnet-4-20250514", messages=messages, tools=tools)
        for block in response.content:
            if block.type == "tool_use":
                result = handle(block.name, block.input)
    """
    tool_defs = [
        {
            "name": t["name"],
            "description": t["description"],
            "input_schema": t["parameters"],
        }
        for t in TOOLS
    ]

    def handle_tool_call(name: str, args: dict[str, Any]) -> dict:
        return call_tool(fs, name, args)

    return tool_defs, handle_tool_call
